x/wasm/ibc.go

```diff
@@ -326,22 +333,142 @@ func (i IBCHandler) OnTimeoutPacket(ctx sdk.Context, packet channeltypes.Packet,
+// see https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-008-app-caller-cbs.md#contractkeeper
+func (i IBCHandler) IBCOnAcknowledgementPacketCallback(
+	cachedCtx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+	contractAddress,
+	packetSenderAddress string,
+) error {
+	contractAddr, err := validateSender(contractAddress, packetSenderAddress)
+	if err != nil {
+		return err
+	}
+
+	msg := wasmvmtypes.IBCSourceCallbackMsg{
+		Acknowledgement: &wasmvmtypes.IBCAckCallbackMsg{
+			Acknowledgement: wasmvmtypes.IBCAcknowledgement{Data: acknowledgement},
+			OriginalPacket:  newIBCPacket(packet),
+			Relayer:         relayer.String(),
+		},
+	}
+	err = i.keeper.IBCSourceCallback(cachedCtx, contractAddr, msg)
+	if err != nil {
+		return errorsmod.Wrap(err, "on source chain callback ack")
+	}
+
+	return nil
+}
+
+// IBCOnTimeoutPacketCallback implements the IBC Callbacks ContractKeeper interface
+// see https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-008-app-caller-cbs.md#contractkeeper
+func (i IBCHandler) IBCOnTimeoutPacketCallback(
+	cachedCtx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+	contractAddress,
+	packetSenderAddress string,
+) error {
+	contractAddr, err := validateSender(contractAddress, packetSenderAddress)
+	if err != nil {
+		return err
+	}
+
+	msg := wasmvmtypes.IBCSourceCallbackMsg{
+		Timeout: &wasmvmtypes.IBCTimeoutCallbackMsg{
+			Packet:  newIBCPacket(packet),
+			Relayer: relayer.String(),
+		},
+	}
+	err = i.keeper.IBCSourceCallback(cachedCtx, contractAddr, msg)
+	if err != nil {
+		return errorsmod.Wrap(err, "on source chain callback timeout")
+	}
+	return nil
+}
+
+// IBCReceivePacketCallback implements the IBC Callbacks ContractKeeper interface
+// see https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-008-app-caller-cbs.md#contractkeeper
+func (i IBCHandler) IBCReceivePacketCallback(
+	cachedCtx sdk.Context,
+	packet ibcexported.PacketI,
+	ack ibcexported.Acknowledgement,
+	contractAddress string,
+) error {
+	// sender validation makes no sense here, as the receiver is never the sender
+	contractAddr, err := sdk.AccAddressFromBech32(contractAddress)
+	if err != nil {
+		return err
+	}
+
+	msg := wasmvmtypes.IBCDestinationCallbackMsg{
+		Ack:    wasmvmtypes.IBCAcknowledgement{Data: ack.Acknowledgement()},
+		Packet: newIBCPacket(packet),
+	}
+
+	err = i.keeper.IBCDestinationCallback(cachedCtx, contractAddr, msg)
+	if err != nil {
+		return errorsmod.Wrap(err, "on destination chain callback")
+	}
+
+	return nil
+}
+
+func validateSender(contractAddr, senderAddr string) (sdk.AccAddress, error) {
+	// We only allow the contract that sent the message to receive source chain callbacks for it.
+	if !strings.EqualFold(contractAddr, senderAddr) {
+		return nil, errorsmod.Wrapf(types.ErrExecuteFailed, "contract address %s does not match packet sender %s", contractAddr, senderAddr)
+	}
```

Comment:
Looks correct, but if we can avoid it I'd not do string comparisions on bech32 inputs. Can we just decode both and then compare the bytes?